File: ahorcado.py
from pyswip import Prolog

# Inicializar el motor Prolog
prolog = Prolog()
prolog.consult("ahorcado.pl")  # Reemplaza con la ruta a tu archivo Prolog

def display_word(word, guessed_letters):
    displayed = [letter if letter in guessed_letters else '_' for letter in word]
    return ' '.join(displayed)

def start_game():
    prolog.query("start.")
    word = list(prolog.query("in_mind(Word), guess(Word, [])."))[0]['Word']
    guessed_letters = []

    while '_' in display_word(word, guessed_letters):
        print(display_word(word, guessed_letters))
        guess = input("Next letter: ").lower()

        if len(guess) == 1 and guess.isalpha():
            result = list(prolog.query(f"guess(Word, {guessed_letters + [guess]})."))[0]
            guessed_letters = result['GuessedLetters']
            remaining_word = result['Word']
        else:
            print("Invalid input. Please enter a single letter.")

    print(f"Congratulations! The word is {''.join(word)}")

if __name__ == "__main__":
    start_game()

File: ahorcado.pl
in_mind([l, o, v, e, l]).


start :- 
    in_mind(Word),
    guess(Word, []).

guess([], _) :- 
    write('Congratulations! The word is '), 
    in_mind(W), 
    write(W), 
    !.


guess(Word, GuessedLetters) :- 
    repeat, in_mind(CorrectWord),
    display_word(CorrectWord, GuessedLetters),
    write('Next letter: '), 
    read(X),
    (   member(X, Word),
        (   \+ member(X, GuessedLetters)  % Verificar que la letra no se haya adivinado antes
    ->  write('OK- ')), 
        append(GuessedLetters, [X], UpdatedGuessedLetters),
        delete(Word, X, RemainingWord),
        guess(RemainingWord, UpdatedGuessedLetters)
    ;   write('Fail. Try again! '), guess(Word, GuessedLetters)
    ).

% Predicado para mostrar la palabra con letras adivinadas y guiones bajos (_) para letras no adivinadas
display_word(Word, CurrentWord) :-
    display_helper(Word, CurrentWord).

display_helper([], _).
display_helper([H | T], CurrentWord) :-
    ( member(H, CurrentWord)
    ->  write(H)
    ;   write('_')
    ),
    display_helper(T, CurrentWord).

print_list([]).
print_list([H | T]) :-
    write(H),
    print_list(T).
    

% Inicializar el sistema con start/0
:- initialization(start).
